File: Android/_setup/_ProjectName_/src/main/java/com/twist/dataengine/entities/TM_FilterAttribute.java
package com.twist.dataengine.entities;

import com.utils.DataHelper;

import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

public class TM_FilterAttribute {

    public String attribute;
    public String query_type;
    //public String slugs;
    //public String taxo;
    //public int position;
    //public boolean visible;
    //public boolean variation;

    public List<TM_FilterAttributeOption> options = new ArrayList<>();

    public TM_FilterAttributeOption getWithSlug(String slug) {
        for (TM_FilterAttributeOption option : options) {
            if (option.slug.equalsIgnoreCase(slug))
                return option;
        }
        return null;
    }

    public TM_FilterAttributeOption getWithName(String name) {
        for (TM_FilterAttributeOption option : options) {
            if (DataHelper.compareAttributeStrings(option.name, name))
                return option;
        }
        return null;
    }

    public TM_FilterAttributeOption getWithName(String name, boolean slugify) {
        for (TM_FilterAttributeOption option : options) {
            if (slugify) {
                if (DataHelper.compareAttributeStrings(option.name, name))
                    return option;
            } else if (option.name.equals(name)) {
                return option;
            }
        }
        return null;
    }

    public boolean isSubsetOf(TM_FilterAttribute other) {
        if (other == null)
            return false;

        if (!this.attribute.equalsIgnoreCase(other.attribute))
            return false;

        if (this.options.size() > other.options.size())
            return false;

        for (TM_FilterAttributeOption option : this.options) {
            if (!other.hasOption(option)) {
                return false;
            }
        }
        return true;
    }


    public boolean isSubsetOf(TM_Attribute other) {
        if (other == null)
            return false;

        if (!this.attribute.equalsIgnoreCase(other.name))
            return false;

        if (this.options.size() > other.options.size())
            return false;

        for (TM_FilterAttributeOption option : this.options) {
            if (!other.hasOptionFilterAttribute(option)) {
                return false;
            }
        }
        return true;
    }

    public boolean hasOption(TM_FilterAttributeOption otherOption) {
        for (TM_FilterAttributeOption option : this.options) {
            if (option.slug.equals(otherOption.slug)) {
                return true;
            }
        }
        return false;
    }


    public boolean removeOption(TM_FilterAttributeOption optionToRemove) {
        for (TM_FilterAttributeOption option : this.options) {
            if (option.slug.equals(optionToRemove.slug)) {
                this.options.remove(option);
                return true;
            }
        }
        return false;
    }

    public boolean hasOptionStr(String optionValue) {
        for (TM_FilterAttributeOption option : this.options) {
            if (option.name.equalsIgnoreCase(optionValue)) {
                return true;
            }
        }
        return false;
    }

    public TM_Attribute getProductAttribute() {
        TM_Attribute attribute = new TM_Attribute(UUID.randomUUID().toString());
        attribute.name = this.attribute;
        if (!this.options.isEmpty()) {
            attribute.slug = this.options.get(0).slug;
            attribute.taxo = this.options.get(0).taxo;
        } else {
            attribute.slug = this.attribute;
            attribute.taxo = "";
        }
        for (TM_FilterAttributeOption option : this.options) {
            attribute.options.add(option.name);
        }
        return attribute;
    }
}

File: IOS/TMStore/TMStore/Classes/Models/TM_FilterAttribute.m
//
//  TM_FilterAttribute.m
//  TMStore
//
//  Created by Rishabh Jain on 16/05/16.
//  Copyright © 2016 Twist Mobile. All rights reserved.
//

#import "TM_FilterAttribute.h"
#import "Utility.h"
@interface TM_FilterAttribute()
@property NSMutableArray* options;//Array of TM_FilterAttributeOption
@end
@implementation TM_FilterAttribute
- (NSMutableArray*)getXYZOptions {
    return _options;
}
- (void)sortOptions {
    if (_options) {
        NSMutableArray* sortedOptions = [[NSMutableArray alloc] init];
        NSArray *sortDescriptors = @[[NSSortDescriptor sortDescriptorWithKey:@"name" ascending:YES]];
        NSArray *sortedArray = [_options sortedArrayUsingDescriptors:sortDescriptors];
        sortedOptions = [[NSMutableArray alloc] initWithArray:sortedArray];
        _options = sortedOptions;
    }
}
- (id)init {
    self = [super init];
    if (self) {
        _attribute = @"";
        _query_type =  @"";
        _title =  @"";
        _display_type =  @"";
        _options = [[NSMutableArray alloc] init];
    }
    return self;
}
- (TM_FilterAttributeOption*)getWithSlug:(NSString*)slug {
    for (TM_FilterAttributeOption* option in _options) {
        if ([Utility compareAttributeNames:option.slug name2:slug]) {
            return option;
        }
    }
    return nil;
}
- (TM_FilterAttributeOption*)getWithName:(NSString*)name {
    for (TM_FilterAttributeOption* option in _options) {
        if ([Utility compareAttributeNames:option.name name2:name]) {
            return option;
        }
    }
    return nil;
}
- (BOOL)isSubsetOf:(TM_FilterAttribute*)other {
    if (other == nil)
        return false;
    
    if (![Utility compareAttributeNames:self.attribute name2:other.attribute]) {
        return false;
    }

    if ((int)[self.options count] > (int)[other.options count]) {
        return false;
    }
    
    for (TM_FilterAttributeOption* option in _options) {
        if (![other hasOption:option]) {
            return false;
        }
    }
    return true;
}
- (BOOL)hasOption:(TM_FilterAttributeOption*)otherOption {
    for (TM_FilterAttributeOption* option in _options) {
        if ([option.slug isEqualToString:otherOption.slug]) {
            return true;
        }
    }
    return false;
}
- (BOOL)isSubsetOfAttribute:(Attribute*)other {
    if (other == nil)
        return false;
    if (![Utility compareAttributeNames:self.attribute name2:other._name]) {
        return false;
    }
    
    if ((int)[self.options count] > (int)[other._options count]) {
        return false;
    }
    
    for (TM_FilterAttributeOption* option in _options) {
        if (![other hasOptionFilterAttribute:option]) {
            return false;
        }
    }
    return true;
}
- (BOOL)removeOption:(TM_FilterAttributeOption*)optionToRemove {
    for (TM_FilterAttributeOption* option in _options) {
        if ([option.slug isEqualToString:optionToRemove.slug]) {
            [_options removeObject:option];
            return true;
        }
    }
    return false;
}
- (BOOL)hasOptionStr:(NSString*)optionValue {
    for (TM_FilterAttributeOption* option in _options) {
        if ([Utility compareAttributeNames:option.name name2:optionValue]) {
            return true;
        }
    }
    return false;
}
- (Attribute*)getProductAttribute {
    Attribute* attribute = [[Attribute alloc] init];
    attribute._name = _attribute;
    if ([self.options count] != 0) {
        attribute._slug = ((TM_FilterAttributeOption*)[self.options objectAtIndex:0]).slug;
        attribute._taxo = ((TM_FilterAttributeOption*)[self.options objectAtIndex:0]).taxo;
    }else{
        attribute._slug = self.attribute;
        attribute._taxo = @"";
    }
    
    for (TM_FilterAttributeOption* option in _options) {
        [attribute._options addObject:option.name];
    }
    return attribute;
}
@end
